### ISSUE TITLE
ansible_connection priority unclear, -c does not override inventory host vars

### ISSUE THREAD

**Issue by IPvSean** [ansible/ansible#50162]
##### SUMMARY
I want to preface this with I want to have this conversation because I am finding it really difficult to create multi-transport playbooks (this group runs on netconf, this group uses httpapi, this group uses network_cli) and not intuitive.  Some of this might end up as documentation but I really want to see what others think and track this.

There is a variety of ways you can set a host connection type.  You can set it in the playbook, the task level, the inventory, or group_vars or host_vars and then if you are using variables, it follows the variable precedence schema which does make sense and is well documented.  It seems odd that the playbook, task, and command line argument are lower than the variable precedence.  That seems intuitive 

Basically the connection precedence is->
- variable (any type), so in my case inventory always wins unless i use extra_vars
- task
- play
- command line `-c` argument

what I think it _should_ be is this:
- command line `-c` argument
- task
- play
- variable (any type)

##### ISSUE TYPE
- Documentation Report

##### COMPONENT NAME
ansible_connection

##### ANSIBLE VERSION

```paste below
[student2@ansible 4-0-tower-setup]$ ansible --version
ansible 2.7.0
  config file = /home/student2/.ansible.cfg
  configured module search path = [u'/home/student2/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Sep 12 2018, 05:31:16) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
```

##### CONFIGURATION

```paste below
[student2@ansible 4-0-tower-setup]$ ansible-config dump --only-changed
DEFAULT_HOST_LIST(/home/student2/.ansible.cfg) = [u'/home/student2/networking-worksho
DEFAULT_TIMEOUT(/home/student2/.ansible.cfg) = 60
DEPRECATION_WARNINGS(/home/student2/.ansible.cfg) = False
HOST_KEY_CHECKING(/home/student2/.ansible.cfg) = False
```

##### OS / ENVIRONMENT
N/A, but RHEL 7.4

##### STEPS TO REPRODUCE
My inventory
```
[student2@ansible 4-0-tower-setup]$ cat ../../lab_inventory/hosts
[all:vars]
ansible_port=22

[routers:children]
cisco
juniper

[routers:vars]
ansible_ssh_private_key_file=/home/student2/.ssh/aws-private.pem

[cisco]
rtr1 ansible_host=35.183.5.74 private_ip=172.16.87.78
rtr2 ansible_host=35.182.224.186 private_ip=172.17.174.65

[juniper]
rtr3 ansible_host=35.182.169.23 private_ip=172.16.35.85
rtr4 ansible_host=35.183.177.32 private_ip=172.17.168.39

[cisco:vars]
ansible_user=ec2-user
ansible_network_os=ios
ansible_connection=network_cli

[juniper:vars]
ansible_user=jnpr
ansible_network_os=junos
ansible_connection=netconf

[dc1]
rtr1
rtr3

[dc2]
rtr2
rtr4

[hosts]
host1 ansible_host=35.183.177.166 ansible_user=ec2-user private_ip=172.17.32.56

[control]
ansible ansible_host=35.182.87.144 ansible_user=ec2-user private_ip=172.16.52.12
```

ad-hoc playbook with `-c network_cli` gets overridden by the inventory
```yaml
[student2@ansible 4-0-tower-setup]$ ansible -m junos_netconf juniper -c network_cli
rtr4 | FAILED! => {
    "changed": false,
    "msg": "Connection type 'netconf' is not valid for 'junos_netconf' module. Please see https://docs.ansible.com/ansible/latest/network/user_guide/platform_junos.html"
}
rtr3 | FAILED! => {
    "changed": false,
    "msg": "Connection type 'netconf' is not valid for 'junos_netconf' module. Please see https://docs.ansible.com/ansible/latest/network/user_guide/platform_junos.html"
}
```

If I try to make a playbook it also fails
```
[student2@ansible 4-0-tower-setup]$ cat netconf_setup.yml
---
- name: TURN ON NETCONF
  hosts: juniper
  gather_facts: false
  connection: network_cli
  tasks:
    - name: TURN ON NETCONF
      junos_netconf:
      connection: network_cli
```

the output
```
[student2@ansible 4-0-tower-setup]$ ansible-playbook netconf_setup.yml

PLAY [TURN ON NETCONF] **************************************************************

TASK [TURN ON NETCONF] **************************************************************
fatal: [rtr4]: FAILED! => {"changed": false, "msg": "Connection type 'netconf' is not valid for 'junos_netconf' module. Please see https://docs.ansible.com/ansible/latest/network/user_guide/platform_junos.html"}
fatal: [rtr3]: FAILED! => {"changed": false, "msg": "Connection type 'netconf' is not valid for 'junos_netconf' module. Please see https://docs.ansible.com/ansible/latest/network/user_guide/platform_junos.html"}
	to retry, use: --limit @/home/student2/networking-workshop/exercises/4-0-tower-setup/netconf_setup.retry

PLAY RECAP **************************************************************************
rtr3                       : ok=0    changed=0    unreachable=0    failed=1
rtr4                       : ok=0    changed=0    unreachable=0    failed=1
```

so both the playbook level and task level don't override it either, in fact the only way to override a variable->
```
ansible-playbook netconf_setup.yml -e "ansible_connection=network_cli"
```

then it will work (b/c -e extra_var overrides all other vars)

##### EXPECTED RESULTS
like this->

```
[student2@ansible 4-0-tower-setup]$ ansible-playbook netconf_setup.yml -c network_cli

PLAY [TURN ON NETCONF] **************************************************************

TASK [TURN ON NETCONF] **************************************************************
changed: [rtr3]
changed: [rtr4]

PLAY RECAP **************************************************************************
rtr3                       : ok=1    changed=1    unreachable=0    failed=0
rtr4                       : ok=1    changed=1    unreachable=0    failed=0
```

##### ACTUAL RESULTS
already shown above, it fails b/c it thinks its on netconf already, the hostvar in inventory always wins unless i override it

**Comment by bcoca** [comment]
List Information
================

Hi!

Thanks very much for your interest in Ansible.  It sincerely means a lot to us. 

This appears to be a user question, and we'd like to direct these kinds of things to either the mailing list or the IRC channel.

   * IRC: #ansible on irc.freenode.net   
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

If you can stop by there, we'd appreciate it.  This allows us to keep the issue tracker for bugs, pull requests, RFEs and the like.

Thank you once again and we look forward to seeing you on the list or IRC.  Thanks!

In any case, this can be quite complex topic, the actual precedence:

`config < cli  options < playbook < 'connection variable'`

There are many reasons for this, the main one has to do with heterogeneous environments, in which a host or group of hosts will have different 'connection' information than the others and should always be accessed in this way. You are not mandated to use the variables, but most people do out of convenience and then hit the 'inconvenient variance' as you just did.

Now some of these can also be split:
```
config: config file < env var
    config file: /etc/ansible/ansible.cfg < ~/.ansible.cfg < `pwd`ansible.cfg
    env var: last defined
```

```
playbook:
  play < [roles/block] < task
```

`variables`: ... long list here https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable

Changing this will break many playbooks as this has always been the case for Ansible and people rely on this behaviour. 

To do what you want, you can either switch away from  variables OR use variable precedence in your favor:

```yaml
- name: TURN ON NETCONF
  hosts: juniper
  gather_facts: false
  tasks:
    - name: TURN ON NETCONF
      junos_netconf:
      vars:
         ansible_connection: network_cli
```

This system might be counter intuitive some times but allows for the most flexibility when dealing with the different environments users are forced to tackle.

**Comment by jctanner** [comment]
reopening as a docs bug.

**Comment by abenokraitis** [comment]
At the very least this should still be tracked and therefore not be closed (since this is a valid issue).

**Comment by abenokraitis** [comment]
Regardless if this is a docs bug or not, seems like users would still naturally want to do:
`ansible-playbook netconf_setup.yml -c network_cli` (doesn't work)
instead of 
`ansible-playbook netconf_setup.yml -e “ansible_connection=network_cli”` (works)

**Comment by bcoca** [comment]
It DOES work, depending on context, here you are assuming `ansible_connection` is set in inventory, that is not always the case.

**Comment by samdoran** [comment]
This behavior has always been the case. We need to document this as well as add tests to ensure this behavior does not change.

**Comment by acozine** [comment]
@abenokraitis @IPvSean is this issue addressed by #50201, which created https://docs.ansible.com/ansible/devel/reference_appendices/general_precedence.html?

**Comment by sivel** [comment]
Closing as resolved by #50201
